lotman/lotman.go

```diff
@@ -0,0 +1,655 @@
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you
+* may not use this file except in compliance with the License.  You may
+* obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+***************************************************************/
+
+package lotman
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"runtime"
+	"sync"
+	"unsafe"
+
+	"github.com/ebitengine/purego"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/pelicanplatform/pelican/param"
+)
+
+var (
+	// A mutex for the Lotman caller context -- make sure we're calling lotman functions with the appropriate caller
+	callerMutex = sync.RWMutex{}
+
+	// Global vars used for one-time Lotman lib initialization
+	lotmanInitTried   = false
+	lotmanInitSuccess = false
+
+	initializedLots []Lot
+
+	// Lotman func signatures we'll bind to the underlying C headers
+	LotmanVersion func() string
+	// Strings in go are immutable, so they're actually passed to the underlying SO as `const`. To get dynamic
+	// output, we need to pass a pointer to a byte array
+	LotmanAddLot              func(lotJSON string, errMsg *[]byte) int32
+	LotmanGetLotJSON          func(lotName string, recursive bool, output *[]byte, errMsg *[]byte) int32
+	LotmanAddToLot            func(additionsJSON string, errMsg *[]byte) int32
+	LotmanUpdateLot           func(updateJSON string, errMsg *[]byte) int32
+	LotmanDeleteLotsRecursive func(lotName string, errMsg *[]byte) int32
+
+	// Auxilliary functions
+	LotmanLotExists     func(lotName string, errMsg *[]byte) int32
+	LotmanSetContextStr func(contextKey string, contextValue string, errMsg *[]byte) int32
+	// Functions that would normally take a char *** as an argument take an *unsafe.Pointer instead because
+	// these functions are responsible for allocating and deallocating the memory for the char ***. The Go
+	// runtime will handle the memory management for the *unsafe.Pointer.
+	LotmanGetLotOwners func(lotName string, recursive bool, output *unsafe.Pointer, errMsg *[]byte) int32
+	// Here, getSelf means get the lot proper if it's a self parent
+	LotmanGetLotParents  func(lotName string, recursive bool, getSelf bool, output *unsafe.Pointer, errMsg *[]byte) int32
+	LotmanGetLotsFromDir func(dir string, recursive bool, output *unsafe.Pointer, errMsg *[]byte) int32
+)
+
+type (
+	Int64FromFloat struct {
+		Value int64
+	}
+
+	LotPaths struct {
+		Path      string `json:"path" mapstructure:"Path"`
+		Recursive bool   `json:"recursive" mapstructure:"Recursive"`
+		LotName   string `json:"lot_name,omitempty"`
```

Comment:
Can `omitempty` be changed to `omit_empty` for consistency in json value names.

Reply:
This can't be changed because `omitempty` is a special flag that prevents the JSON from injecting a key that would otherwise have a null value. See the [here](https://pkg.go.dev/encoding/json).